python/dgl/contrib/dis_kvstore.py

```diff
@@ -220,72 +250,84 @@ def _push_handler(self, name, ID, data):
+        target : dict of data
+            self._data_store
         """
-        for idx in range(ID.shape[0]):
-            self._data_store[name][ID[idx]] += data[idx]
+        target[name][ID] += data
 
-    def _pull_handler(self, name, ID):
+    def _pull_handler(self, name, ID, target):
         """Default handler for PULL operation.
 
-        On default, _pull_handler perform gather_row() operation for the tensor.
+        On default, _pull_handler perform index select operation for the tensor.
 
         Parameters
         ----------
         name : str
             data name
         ID : tensor (mx.ndarray or torch.tensor)
-            a vector storing the IDs that has been re-mapped to local id.
+            a vector storing the ID list.
+        target : dict of data
+            self._data_store
 
         Return
         ------
         tensor
-            a tensor with the same row size of ID
+            a tensor with the same row size of ID.
         """
-        new_tensor = F.gather_row(self._data_store[name], ID)
-        return new_tensor
+        return target[name][ID]
 
 class KVClient(object):
     """KVClient is used to push/pull tensors to/from KVServer on DGL trainer.
 
-    There are five operations supported by KVClient:
-
-      * init_data(name, server_id, shape, init_type, low, high): 
-          initialize tensor on target KVServer.
-      * push(name, server_id, id_tensor, data_tensor): 
-          push sparse data to KVServer given specified ID.
-      * pull(name, server_id, id_tensor): 
-          pull sparse data from KVServer given specified ID.
-      * pull_wait(): 
-          wait scheduled pull operation finish its job.
-      * shut_down(): 
-          shut down all KVServer nodes.
+    Note that, DO NOT share one KVClient in multiple threads in python because this behavior is not defined.
 
-    Note that, DO NOT use KVClient in multiple threads!
+    For now, KVClient can only run in CPU, we will support GPU KVClient in the future.
 
     Parameters
     ----------
     client_id : int
         KVClient's ID (start from 0)
+    local_server_id : int
+        ID of server that located in same node with current client. By this ID, client knows that when performing local
+        update through shared-memory instead of using tcp network.
     server_namebook: dict
         IP address namebook of KVServer, where key is the KVServer's ID 
-        (start from 0) and value is the server's IP address, e.g.,
+        (start from 0) and value is the server's IP address and port, e.g.,
 
         { 0:'168.12.23.45:50051', 
           1:'168.12.23.21:50051', 
           2:'168.12.46.12:50051' }
     client_addr : str
-        IP address of current KVClient, e.g., '168.12.23.22:50051'
+        IP address and port of current KVClient, e.g., '168.12.23.22:50051'
+    partition_book : list or tensor (mx.ndarray or torch.tensor)
+        A book that maps global ID to target server ID.
```

Comment:
if we don't shuffle vertices after metis partitioning, every machine needs to maintain a full map from global vid to their locations. it seems pretty expensive.